projects/components/src/code-viewer/code-viewer.component.ts

```diff
@@ -0,0 +1,110 @@
+import { ChangeDetectionStrategy, Component, Input, OnChanges, TemplateRef } from '@angular/core';
+import { Color, TypedSimpleChanges } from '@hypertrace/common';
+import { isEmpty, isNil } from 'lodash-es';
+import { DownloadFileMetadata } from '../download-file/download-file-metadata';
+
+@Component({
+  selector: 'ht-code-viewer',
+  styleUrls: ['./code-viewer.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  template: `
+    <div class="code-viewer" [style.backgroundColor]="this.backgroundColor">
+      <div class="header">
+        <div class="title">{{ this.titleText }}</div>
+        <div class="header-content">
+          <ng-container *ngIf="this.additionalHeaderContent">
+            <ng-container *ngTemplateOutlet="this.additionalHeaderContent"></ng-container
+          ></ng-container>
+          <ht-search-box
+            class="search-box"
+            backgroundColor="${Color.White}"
+            (valueChange)="this.onSearch($event)"
+          ></ht-search-box>
+          <ht-download-file *ngIf="this.downloadCodeMetadata" [metadata]="this.downloadCodeMetadata"></ht-download-file>
+        </div>
+      </div>
+      <div class="content">
+        <div class="line-numbers">
+          <div
+            *ngFor="let lineNumber of this.lineNumbers"
+            class="line-number"
+            [ngClass]="{ highlight: this.isHighlighted(lineNumber - 1) }"
+          >
+            {{ lineNumber }}
+          </div>
+        </div>
+        <div class="code-lines">
+          <div
+            *ngFor="let codeLine of this.code; let index = index"
+            class="code-line"
+            [ngClass]="{ highlight: this.isHighlighted(index) }"
+          >
+            <pre>{{ codeLine }}</pre>
+          </div>
+        </div>
+      </div>
+    </div>
+  `
+})
+export class CodeViewerComponent implements OnChanges {
+  @Input()
+  public code: string[] = []; // Pre-formatted code lines as string
+
+  @Input()
+  public highlightText: string = ''; // To highlight the entire line
+
+  @Input()
+  public titleText: string = 'Code Viewer';
+
+  @Input()
+  public backgroundColor: string = Color.OffWhite;
+
+  @Input()
+  public downloadCodeMetadata?: DownloadFileMetadata;
+
+  @Input()
+  public additionalHeaderContent?: TemplateRef<unknown>;
+
+  private codeLineElements?: HTMLElement[];
+
+  public ngOnChanges(changes: TypedSimpleChanges<this>): void {
+    if (changes.code) {
+      this.codeLineElements = undefined;
```

Comment:
hmm.. why don't we create all elements and line numbers, etc right here? Setting it to undefined and then doing it via html seems avoidable

Reply:
This is remove since used `htHighlight` pipe now